oeps/oep-0018-bp-python-dependencies.rst

```diff
@@ -0,0 +1,350 @@
++-----------------+--------------------------------------------------------+
+| Title           | Python Dependencies Management                         |
++-----------------+--------------------------------------------------------+
+| Last Modified   | 2018-03-27                                             |
++-----------------+--------------------------------------------------------+
+| Authors         | Jeremy Bowman <jbowman@edx.org>                        |
++-----------------+--------------------------------------------------------+
+| Arbiter         | Nimisha Asthagiri <nimisha@edx.org>                    |
++-----------------+--------------------------------------------------------+
+| Status          | Draft                                                  |
++-----------------+--------------------------------------------------------+
+| Type            | Best Practice                                          |
++-----------------+--------------------------------------------------------+
+| Created         | 2018-03-27                                             |
++-----------------+--------------------------------------------------------+
+| `Review Period` | 2018-03-27 - 2018-04-20                                |
++-----------------+--------------------------------------------------------+
+| `Resolution`    |                                                        |
++-----------------+--------------------------------------------------------+
+
+Abstract
+========
+
+Proposes best practices for declaring and maintaining dependencies on other
+Python packages in Open edX software repositories.
+
+Motivation
+==========
+
+The Open edX project includes dozens of Python software repositories, most of
+which depend on certain other Python packages being installed in order to
+function correctly.  The simple methods we originally used to do this have
+assorted drawbacks that have repeatedly caused problems over the past few
+years: accidental upgrades to incompatible versions, strict installation
+requirements that restrict the ability of downstream packages to manage their
+own dependency versions, lack of clarity regarding the full set of packages
+actually depended upon, etc.
+
+Outlined here is a recommended standard for declaring dependencies on other
+Python packages which resolves most of these issues and will let us make all
+the Open edX Python packages consistent with each other (and many other open
+source Python projects) for ease of understanding and maintenance.
+
+Specification
+=============
+
+The key to successful Python dependency management is to break it down into
+four parts:
+
+1. Identify the different contexts in which dependencies will need to be
+   installed.
+2. For each of these contexts, declare the direct dependencies that will be
+   needed.  Use the least restrictive constraints which should allow pip to
+   install a working set of dependencies.
```

Comment:
I think I missed a key point here:  our philosophy now is to not explicitly pin versions, and instead to let pip-tools pin them for us.  This is a big shift from how we used to do things, and needs to be highlighted much more strongly.

Reply:
Sure, rewording to try to make things like that more clear.